eng/scripts/Get-BinarySizes.ps1

```diff
@@ -0,0 +1,179 @@
+        return "x86"
+    }
+
+    if ($env:CMAKE_GENERATOR_PLATFORM -and $env:CMAKE_GENERATOR_PLATFORM -eq "x64") {
+        return "x64"
+    }
+
+    # Most builds target x64 by default
+    return "x64"
+}
+
+function getToolChain {
+    if ($OSVmImage.StartsWith('macOS')) {
+        return "AppleClang 12"
+    }
+
+    if ($OSVmImage -eq "MMS2019") {
+        return "MSVC"
+    }
+
+    if ($OSVmImage.Contains("Ubuntu")) {
+        if ($CmakeEnvArg.Contains('g++-5')) {
+            return 'g++-5'
+        } elseif ($env:CXX -and $env:CXX.Contains("g++8")) {
+            return 'g++-8'
+        } elseif ($env:CXX -and $env:CXX.Contains("g++-9")) {
+            return 'g++-9'
+        } elseif ($env:CXX -and $env:CXX.Contains("clang-11")) {
+            return 'clang-11'
+        }
+        return "g++-7"
+    }
+    LogError "Could not infer toolchain"
+}
+
+function getTargetPlatform {
+    if ($OSVmImage.StartsWith('macOS')) {
+        return "macos"
+    }
+
+    if ($OSVmImage -eq 'MMS2019') {
+        if (!$env:CMAKE_SYSTEM_NAME -and !$CmakeArgs.Contains('WindowsStore')) {
+            return 'win32'
+        } elseif ($env:CMAKE_SYSTEM_NAME -eq 'WindowsStore' -or $CmakeArgs.Contains('WindowsStore')) {
+            return 'uwp'
+        }
+    }
+
+    if ($OSVmImage.Contains("Ubuntu")) {
+        return 'linux'
+    }
+
+    LogError "Could not infer target platform"
+}
+
+function getBuildConfiguration {
+    if ($env:CMAKE_BUILD_TYPE) {
+        return $env:CMAKE_BUILD_TYPE
+    }
+
+    if ($BuildArgs.Contains("--config Debug")) {
+        return "Debug"
+    }
+
+    if ($BuildArgs.Contains("--config Release")) {
+        return "Release"
+    }
+
+    # Most builds default to Debug unless overridden by configuration variables
+    return "Debug"
+}
+
+$searchPath = "$BuildDirectory/sdk/$ServiceDirectory/*/*.a"
+if ($IsWindows) {
+    $searchPath = "$BuildDirectory/sdk/$ServiceDirectory/*/*/*.lib"
+}
+
+$binaries = Get-ChildItem -Path $searchPath
+
+$buildLabels = @{
+    TargetOs = getTargetOs;
+    Toolchain = getToolChain;
+    BuildConfig = getBuildConfiguration;
+    TargetArchitecture = getTargetArchitecture;
+    TargetPlatform = getTargetPlatform;
+}
+
+if ($CI) {
+    foreach ($binary in $binaries) {
+        $metricLogObject = @{
+            name = "BinarySize";
+            value = $binary.Length
+            timestamp = (Get-Date -AsUTC).ToString("o")
+            labels = @{
+                Job = $Job;
+                BuildReason = $BuildReason;
+                SourceBranch = $SourceBranch;
+                BinaryName = $binary.Name;
+            } + $buildLabels + $ExtraLabels
+        }
+
+        $metricLogJson = ConvertTo-Json $metricLogObject -Compress
+        Write-Host "logmetric: $metricLogJson"
+    }
+}
+$binaries `
+    | Format-Table -Property Name, @{Name="SizeInKB"; Expression={"{0:N2}" -f ($_.Length / 1KB)}; Alignment='right'} `
```

Comment:
Directory size isn't always the best way of measuring binary size, it might be interesting to use `link /dump` and `readelf -S` to dump the individual elements.

Also, C++ template expansion is notorious for bloating libraries, comdat folding eliminates almost all of that bloat so lib size isn't always a decent metric for "size".

Reply:
These are the sizes of the individual files, not directories... which doesn't alter the validity of your point that static linked library files do not give an accurate picture of the ultimate size of a program which uses the library. 

We're already publishing the `.map` files based on our test executables into artifacts for a more fine-grained view of binary sizes... however map file formats are specific to the linker used and so we don't have an easy way to compare or report metrics on sizes across the various platforms yet.